enhancements/network/external-lb-vips.md

```diff
@@ -0,0 +1,212 @@
+We will make some change in `install-config.yaml` so customer can configure the
+deployment to use an external loadbalancer if they want.
+Then the components which interact with networking will have to check this
+new field and handle this scenario appropriately.
+
+### Workflow Description
+
+1. The customer will configure their external loadbalancer(s), which includes configuring the VIPs
+   for API and Ingress.
+2. The customer will populate their external DNS server with required records for api, api-int, and ingress.
+3. The customer will write the `install-config.yaml` file and set the new `loadBalancerType` field to `external`, and
+   will not set the fields for apiVIP and ingressVIP.
+4. In the installer, if the platform does support external loadbalancer, some validation will be run
+   to check that the VIPs can already be resolved (the validation already exists for vsphere, we need to
+    make it re-usable). A new field in `PlatformStatus` will be set.
+5. In machine-config-operator, if LB type is external, the internal loadbalancer will not be
+   deployed.
+6. When using an external loadbalancer, the VIPs fields won't be set in `PlatformStatus` and the
+   OpenShift components won't use it when they're not present and LB type is external.
+7. After the deployment, the IP address of a service (e.g. API) can change, as long as the external
+   loadbalancer is updated and the DNS record has been updated and the results are propagated as well.
+
+### API Extensions
+
+A new field will be added to the `PlatformStatus` of the infrastructure object to tell
+the kind of the loadbalancer.
+
+```golang
+// LoadBalancerType can be either external or internal 
+type LoadBalancerType string
+const (
+	// LoadBalancerTypeExternal
+	LoadBalancerTypeExternal LoadBalancerType = "external"
+	// LoadBalancerTypeInternal
+	LoadBalancerTypeInternal LoadBalancerType = "internal"
+)
+// VSpherePlatformStatus holds the current status of the vSphere infrastructure provider.
+type VSpherePlatformStatus struct {
+  // loadBalancerType defines if a internal or external load balancer
+  // is to be used for the openshift api and ingress endpoints.
+  // When set to Internal the static pods defined in machine config operator
+  // will be used.
+  // When set to External the static pods will not be deployed. It will be 
+  // expected that the LB and DNS A records be pre-configured prior to installation.
+  // +kubebuilder:default=internal
+  // +kubebuilder:validation:Enum=internal;external
+  LoadBalancer LoadBalancerType `json:"loadBalancerType,omitempty"`
+}
+```
+
+This field will be set in the `install-config.yaml`:
+```yaml
+platform:
+  vsphere:
+    loadBalancerType: external
+```
+
+### Implementation Details/Notes/Constraints
+
+* CoreDNS will not be deployed if external loadbalancer is being used.
+  The customer is responsible of populating the DNS records for api, api-int and ingress.
+  We need to check who is responsible for machine records.
```

Comment:
I wonder if we could allow the user to give records for them (in a separate enhancement). This is for edge compact deployments

Reply:
How would it be done (in your mind)? Via a new API?

Reply:
Why we can't configure coredns? I assume we can create sort of api that will configure coredns to point to api vip, will help a lot i assume

Reply:
I would like to have an API to update coredns, but the drawback there is we don't currently have anything like that. All of the Infrastructure fields are in the Status object, which is not user-modifiable. We'd have to design a whole new mechanism for updating those values (probably the Infrastructure Spec).

I'd still like to do it because it would allow users to keep the internal DNS infrastructure and only have to manage the loadbalancer themselves, but I do realize that could add a significant amount of time to the implementation. It is possible we could do both in parallel. They're not mutually exclusive options, although we'd want to coordinate to make sure we don't step on each other's changes.

Reply:
I agree with Ben and it should indeed be a separate effort. I personally won't have the bandwidth to look at it but would be happy to collaborate so we don't step on each others.

Reply:
We may not necessarily need to update coredns, but just permit setting the API and Ingress records like we already do with the values coming from the VIPs fields in `PlatformStatus`?
I understand vsphere currently uses empty values to detect if it needs to use an external LB. With the introduction of a new `LoadBalancerType` as suggested here, we could rely on this rather than on the values of `IngressIPs` and `APIServerInternalIPs` and then using an external LB would not force you to use an externalize DNS as well.

Reply:
I like this idea @mandre . I'll wait for more comments before updating the text but I think we could do that.
@rvanderp3 @jcpowermac - any feedback on this?

Reply:
Keeping CoreDNS only benefit for in the case of vSphere is only api-int, not sure that makes sense to me.
(and we randomly see issues with api-int resolving so I am less inclined keeping it - we aren't properly testing it in CI because of Windows, an A record is created in Route53 for api-int)

Reply:
Here is what I propose:
* If you use an external LB but don't want CoreDNS at all (which seems to be the case of `vsphere`), set `loadBalancerType` to `external` in your `install-config.yaml` and skip `ingressVIPs` + `apiVIPs` parameters. Their fields will be empty in `PlatformStatus`.
* If you use an external LB but you want to keep CoreDNS, set `loadBalancerType` to `external` in your `install-config.yaml` and provide the IPs in `ingressVIPs` + `apiVIPs`. Their fields will be in `PlatformStatus` as usual, and CoreDNS will be deployed and provide records for `api`, `api-int`, `ingress` and `machines`.

Before I update this enhancement with that, can I get :+1: or :-1:  ?

Reply:
I am wondering if we want that complexity? Having time to think about more the ongoing maintenance would be way easier if all the platforms were the same and just use `loadBalancerType` as the switch makes it unambiguous. Its not going to hurt anything with CoreDNS running - though we should probably get the vSphere IPI CI test resolved regardless.

Reply:
Ok then we set `loadBalancerType` to external and provide `ingressVIPs` + `apiVIPs`; CoreDNS will be deployed and provide records for api, api-int, ingress and machines.
For any forthcoming bug, we threat them as bug as can still adjust the design if really needed. WDYT?

Reply:
Worth noting that if we do keep coredns and don't implement any way to update the VIP fields in PlatformStatus then the external LB address can never change. I know we already enforce that for the internal LB, but now we're talking about things outside of our control. I could see that being a problem in some environments, particularly if the external LB is not specifically dedicated to the OpenShift cluster. At the very least that should be called out as a risk.

Reply:
I'll add that as a risk.
I think that adding the capabilities to update the VIPs in PlatformStatus (and therefore update CoreDNS, which AFIK would be the only thing to update in the cluster?) needs its own enhancement or can be done after this effort. Please :+1: or :-1: . Thanks